test/tokenbridge.test.ts

```diff
@@ -0,0 +1,204 @@
+  let mockStarknetMessaging: Contract;
+  let L1TokenFactory: ContractFactory;
+  let l1tokenA: Contract;
+  let l1tokenB: Contract;
+  let TokenBridgeL1: ContractFactory;
+  let tokenBridgeL1: Contract;
+  let messagingContractAddress: string;
+
+  before(async function () {
+
+    // L2 deployments
+
+    l2user = await starknet.deployAccount("OpenZeppelin");
+
+    L2TokenFactory = await starknet.getContractFactory('l2_token');
+    l2tokenA = await L2TokenFactory.deploy(
+        { name: 1234, symbol: 123, decimals: 18, minter_address: BigInt(l2user.starknetContract.address) });
+    l2tokenB = await L2TokenFactory.deploy(
+        { name: 4321, symbol: 321, decimals: 18, minter_address: BigInt(l2user.starknetContract.address) });
+  
+    TokenBridgeL2 = await starknet.getContractFactory('token_bridge');
+    tokenBridgeL2 = await TokenBridgeL2.deploy({ governor_address: BigInt(l2user.starknetContract.address) });
+
+    // L1 deployments
+
+    [signer, l1user] = await ethers.getSigners();
+
+    MockStarknetMessaging = await ethers.getContractFactory(
+      'MockStarknetMessaging',
+      signer,
+    );
+    mockStarknetMessaging = await MockStarknetMessaging.deploy();
+    await mockStarknetMessaging.deployed();
+
+    L1TokenFactory = await ethers.getContractFactory('L1Token', signer);
+    l1tokenA = await L1TokenFactory.deploy(1000);
+    l1tokenB = await L1TokenFactory.deploy(1000);
+
+    TokenBridgeL1 = await ethers.getContractFactory('TokenBridge', signer);
+    tokenBridgeL1 = await TokenBridgeL1.deploy();
+    await tokenBridgeL1.deployed();
+  });
+
+  it('should deploy the messaging contract', async () => {
+    const {
+      address: deployedTo,
+      l1_provider: L1Provider,
+    } = await starknet.devnet.loadL1MessagingContract(networkUrl);
+    expect(deployedTo).not.to.be.undefined;
+    expect(L1Provider).to.equal(networkUrl);    
+  });
+
+  it('should load the already deployed contract if the address is provided', async () => {
+    const {
+      address: deployedTo,
+    } = await starknet.devnet.loadL1MessagingContract(networkUrl);
+
+    await tokenBridgeL1.initializeWithoutProxy(tokenBridgeL2.address, deployedTo);
+    messagingContractAddress = await tokenBridgeL1.messagingContract();
+
+    const {
+      address: loadedFrom,
+    } = await starknet.devnet.loadL1MessagingContract(
+      networkUrl,
+      messagingContractAddress,
+    );
+    expect(messagingContractAddress).to.equal(loadedFrom);
+  });
+
+  it('should exchange messages between L1 and L2', async () => {
+
+    // on L1: send 200 tokens A and 300 tokens B to l1user
+    await l1tokenA.transfer(l1user.address, 200);
+    await l1tokenB.transfer(l1user.address, 300);
+    expect(await l1tokenA.balanceOf(l1user.address)).to.equal(200);
+    expect(await l1tokenB.balanceOf(l1user.address)).to.equal(300);    
+
+    // load messaging contract
+    await starknet.devnet.loadL1MessagingContract(networkUrl, mockStarknetMessaging.address);
```

Comment:
Can we break out all the initialisation into a before each so we can add tests easily?